React 18 upgrade + API root creation change

wjrosa opened the pull request:
<!--
Did I add a title? A descriptive, yet concise, title.
-->

<!--
Issue: Link to the GitHub issue this PR addresses (if appropriate).
-->

Fixes STRIPE-451
Base PR https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4264

## Changes proposed in this Pull Request:

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for the PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

This PR updates the current React version to 18, along with the related packages. To support it, I am also rewriting the way we create root containers, following the API signature update.

## Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Please follow the following guidelines when writing testing instructions:

- Include screenshots if there is no similar flow in the critical flows: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Critical-flows
- Assume instructions will be copied over to the Release Testing Instructions wiki page: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions
- Assume instructions will be followed by external testers.
- Assume tester does not have intimate knowledge of Stripe.
-->

- Checkout and build this branch on your test environment (`update/react-api-root-creation-change`)
- Check if the frontend tests are still passing
- Confirm the affected screens are correctly rendered:
-- Express checkout settings
-- Main Stripe settings screen
-- WooCommerce payment providers page (`admin.php?page=wc-settings&tab=checkout`)

---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [ ] Tested on mobile (or does not apply)

### Changelog entry

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)
-   [ ] Added the [needs docs label](https://github.com/woocommerce/woocommerce-gateway-stripe/labels?q=docs) (or does not apply)
-   [ ] Included this PR in the Release Thread scope (or does not apply)

wjrosa committed:
React API root creation change

wjrosa committed:
React dom package updates

wjrosa committed:
Merge branch 'develop' into update/react-api-root-creation-change

wjrosa committed:
Fix tests + React 18 upgrade

wjrosa committed:
Fix tests

wjrosa committed:
Adding react-test-renderer package

wjrosa committed:
Reverting unnecessary changes

wjrosa committed:
Changelog and readme entries

daledupreez reviewed (commented):
I am still seeing a lot of console errors when running the tests, including the following:
```
Warning: ReactDOM.render is no longer supported in React 18. Use createRoot instead. Until you switch to the new API, your app will behave as if it's running React 17. Learn more: https://reactjs.org/link/switch-to-createroot
```

```
Warning: unmountComponentAtNode is deprecated and will be removed in the next major release. Switch to the createRoot API. Learn more: https://reactjs.org/link/switch-to-createroot
```

```
Warning: `ReactDOMTestUtils.act` is deprecated in favor of `React.act`. Import `act` from `react` instead of `react-dom/test-utils`. See https://react.dev/warnings/react-dom-test-utils for more info.
```

AFAICT, all of these errors stem from our older version of `@testing-library/react`. where the tests are mostly importing the `render` function from that library.

daledupreez commented:
Also, I got all the tests running without `react-test-renderer` present at all -- I ran `npm uninstall react-test-renderer` and still got the tests running successfully. The update to `react-dom` may have been sufficient.

wjrosa committed:
Removing unnecessary package

wjrosa committed:
Updating @testing-library/react

wjrosa committed:
Replace act function source

wjrosa committed:
Fix tests

wjrosa commented:
@daledupreez good catch! I fixed most of them by updating some other libraries (matching the base PR). The remaining ones are coming indirectly from packages like `@wordpress/components` and `@wordpress/data`. I tried to update those, but that breaks without the Node update.

> Also, I got all the tests running without react-test-renderer present at all -- I ran npm uninstall react-test-renderer and still got the tests running successfully. The update to react-dom may have been sufficient.

Thanks! I removed this package and confirmed everything is working fine now

daledupreez committed:
Merge branch 'develop' into update/react-api-root-creation-change

daledupreez commented:
@wjrosa, I re-tested with your changes and things are looking better, but I was still seeing some console errors. When I dug deeper, I realised that `@testing-library/react-hooks` has mostly been included in `@testing-library/react`, so we can switch the imports to `@testing-library/react` and drop `@testing-library/react-hooks`. I created a branch off this one and got that working in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4619, where I have requested your review, with the goal of merging that PR into this branch.

daledupreez commented:
I also fixed a very small issue in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4620 that wasn't directly touched in this PR, but may stem from stricter checks in React 18. That PR is also using this branch as a target so we can land everything together.

wjrosa commented:
Thanks, Dale! I missed that. I reviewed all your related PRs. Would you mind sharing where you saw that `@testing-library/react-hooks` change? Just to document and out of curiosity.

daledupreez committed:
Remove @testing-library/react-hooks for compatibility with React 18 (#4619)

* Use act and renderHook from @testing-library/react instead of @testing-library/react-hooks
* Remove @testing-library/react-hooks

daledupreez committed:
Fix warning for PaymentMethodDeprecationPill test (#4620)

daledupreez commented:
The docs for `@testing-library/react-hooks` were in the project's GitHub repo: https://github.com/testing-library/react-hooks-testing-library?tab=readme-ov-file#a-note-about-react-18-support
I noticed from inspecting the `renderHooks` implementation in my local install that the `act` and `unmountComponentAt` functions were being called and flagged as deprecated via highlighting.

I have also gone ahead and merged both of my changes into this branch. I'll take another look at the combined tests etc shortly.

daledupreez commented:
Noting for visibility that I am having issues with e2e tests in this branch, both automated and via my local testing -- still digging in to what is going awry.

wjrosa committed:
Merge branch 'develop' into update/react-api-root-creation-change

wjrosa committed:
Merge branch 'develop' into update/react-api-root-creation-change

wjrosa committed:
Fix merge

daledupreez reviewed (commented):
I am seeing some unexpected console errors from the e2e tests that I haven't been able to reproduce directly. The trace data is showing multiple entries with the following error text:

```
TypeError: Cannot read properties of undefined (reading 'createRoot')
    at Object.<anonymous> (http://localhost:8088/wp-content/plugins/woocommerce-gateway-stripe/build/upe-classic.js?ver=9.8.1:1:41219)
    at n (http://localhost:8088/wp-content/plugins/woocommerce-gateway-stripe/build/upe-classic.js?ver=9.8.1:1:110)
    at Object.<anonymous> (http://localhost:8088/wp-content/plugins/woocommerce-gateway-stripe/build/upe-classic.js?ver=9.8.1:1:56319)
    at n (http://localhost:8088/wp-content/plugins/woocommerce-gateway-stripe/build/upe-classic.js?ver=9.8.1:1:110)
    at Object.<anonymous> (http://localhost:8088/wp-content/plugins/woocommerce-gateway-stripe/build/upe-classic.js?ver=9.8.1:1:15488)
    at n (http://localhost:8088/wp-content/plugins/woocommerce-gateway-stripe/build/upe-classic.js?ver=9.8.1:1:110)
    at Object.<anonymous> (http://localhost:8088/wp-content/plugins/woocommerce-gateway-stripe/build/upe-classic.js?ver=9.8.1:1:7155)
    at n (http://localhost:8088/wp-content/plugins/woocommerce-gateway-stripe/build/upe-classic.js?ver=9.8.1:1:110)
    at Object.<anonymous> (http://localhost:8088/wp-content/plugins/woocommerce-gateway-stripe/build/upe-classic.js?ver=9.8.1:1:104783)
    at n (http://localhost:8088/wp-content/plugins/woocommerce-gateway-stripe/build/upe-classic.js?ver=9.8.1:1:110)
```

Based on the timing, this may be a load order issue, but it feels somewhat fishy.

Commit:
Merge branch 'develop' into update/react-api-root-creation-change

annemirasol commented:
I get the errors @daledupreez mentioned for pre-order products and (weirdly) classic checkout:

<img width="764" height="187" alt="Screenshot 2025-08-28 at 12 05 27 PM" src="https://github.com/user-attachments/assets/6e75e775-55ef-484a-b401-03ae4b750711" />

The pre-order e2e test is also consistently failing for my local.

I compared with `develop` and I do not get this error.

wjrosa committed:
Merge branch 'develop' into update/react-api-root-creation-change

wjrosa committed:
Renaming some test files

wjrosa committed:
Fix createRoot console error

wjrosa commented:
I took a look at [the official documentation](https://react.dev/reference/react-dom/createPortal), and it looks like we should use `createPortal` instead of `createRoot` now when not creating the primary root node. This solved the issue for me in `client/stripe-utils/utils.js` (which caused the errors above in the classic checkout).

I was going to change all the other secondary calls to `createRoot` on our code, but I think it is better to wait for the need to arise.

PS: I had to clean my e2e tests environment for them to pass.

daledupreez committed:
Merge branch 'develop' into update/react-api-root-creation-change

annemirasol reviewed (approved):
✅ Verified payment element loading in shortcode checkout for pre-orders

Been getting some pretty consistent failures for certain tests in my local e2e, but they seem to be a Playwright timing issue, as changing `.click()` to `.dispatchEvent( 'click' );` fixes it. I suspect it's a layout shift happening as we click. I also couldn't repro on the browser.

Github workflow e2e runs are green -- LGTM!

daledupreez reviewed (approved):
Given that the e2e tests are green, I think we can get this merged.
